File: sidecar/raylab_sidecar/storage.py
from __future__ import annotations

import json
import os
import sys
from pathlib import Path
from typing import Any

try:
    import keyring
except Exception:  # pragma: no cover - depends on local Python environment
    keyring = None  # type: ignore[assignment]

from .models import AppConfig, AuditEvent, ResourceCaps


APP_NAME = "RayLabClusterManager"
LEGACY_DEFAULT_CAPS = ResourceCaps()


def config_dir() -> Path:
    override = os.environ.get("RAYLAB_CONFIG_DIR")
    if override:
        return Path(override)
    if os.name == "nt":
        base = Path(os.environ.get("APPDATA", Path.home() / "AppData" / "Roaming"))
    elif sys.platform == "darwin":
        base = Path.home() / "Library" / "Application Support"
    else:
        base = Path(os.environ.get("XDG_CONFIG_HOME", Path.home() / ".config"))
    return base / "raylab-cluster-manager"


def config_path() -> Path:
    return config_dir() / "config.json"


class ConfigStore:
    def __init__(self, path: Path | None = None) -> None:
        self.path = path or config_path()

    def load(self) -> AppConfig:
        if not self.path.exists():
            return AppConfig(resource_caps=_detected_resource_caps())
        data = json.loads(self.path.read_text(encoding="utf-8"))
        config = AppConfig.model_validate(data)
        if _uses_legacy_default_caps(config.resource_caps):
            config.resource_caps = _detected_resource_caps()
            self.save(config)
        return config

    def save(self, config: AppConfig) -> AppConfig:
        self.path.parent.mkdir(parents=True, exist_ok=True)
        self.path.write_text(config.model_dump_json(indent=2), encoding="utf-8")
        return config

    def append_audit(self, event: AuditEvent) -> AppConfig:
        config = self.load()
        config.audit.insert(0, event)
        config.audit = config.audit[:1000]
        return self.save(config)


def _uses_legacy_default_caps(caps: ResourceCaps) -> bool:
    return (
        caps.cpus == LEGACY_DEFAULT_CAPS.cpus
        and caps.gpus == LEGACY_DEFAULT_CAPS.gpus
        and caps.memory_gb == LEGACY_DEFAULT_CAPS.memory_gb
        and caps.gpu_memory_gb == LEGACY_DEFAULT_CAPS.gpu_memory_gb
        and caps.max_concurrent_jobs == LEGACY_DEFAULT_CAPS.max_concurrent_jobs
    )


def _detected_resource_caps() -> ResourceCaps:
    try:
        from .hardware import detect_hardware

        hardware = detect_hardware()
    except Exception:
        return LEGACY_DEFAULT_CAPS.model_copy()

    return ResourceCaps(
        cpus=max(1, hardware.cpu_logical or int(LEGACY_DEFAULT_CAPS.cpus)),
        gpus=max(0, hardware.gpu_count),
        memory_gb=max(0.1, hardware.memory_total_gb or LEGACY_DEFAULT_CAPS.memory_gb),
        gpu_memory_gb=max(0, hardware.gpu_memory_total_gb or 0),
        max_concurrent_jobs=LEGACY_DEFAULT_CAPS.max_concurrent_jobs,
    )


class SecretStore:
    def get(self, ref: str) -> str | None:
        fallback = config_dir() / "dev-secrets.json"
        if keyring is None:
            if fallback.exists():
                return json.loads(fallback.read_text(encoding="utf-8")).get(ref)
            return None
        try:
            return keyring.get_password(APP_NAME, ref)
        except Exception:
            if fallback.exists():
                return json.loads(fallback.read_text(encoding="utf-8")).get(ref)
            return None

    def set(self, ref: str, value: str) -> None:
        fallback = config_dir() / "dev-secrets.json"
        if keyring is None:
            fallback.parent.mkdir(parents=True, exist_ok=True)
            data: dict[str, Any] = {}
            if fallback.exists():
                data = json.loads(fallback.read_text(encoding="utf-8"))
            data[ref] = value
            fallback.write_text(json.dumps(data, indent=2), encoding="utf-8")
            return
        try:
            keyring.set_password(APP_NAME, ref, value)
        except Exception:
            fallback.parent.mkdir(parents=True, exist_ok=True)
            data: dict[str, Any] = {}
            if fallback.exists():
                data = json.loads(fallback.read_text(encoding="utf-8"))
            data[ref] = value
            fallback.write_text(json.dumps(data, indent=2), encoding="utf-8")

    def get_or_create_token(self, ref: str, length: int = 48) -> str:
        import secrets

        existing = self.get(ref)
        if existing:
            return existing
        token = secrets.token_urlsafe(length)
        self.set(ref, token)
        return token

File: electron/src/storage.js
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const crypto = require('crypto');

// ─── Config directory ─────────────────────────────────────────────────────────

function configDir() {
  if (process.env.RAYLAB_CONFIG_DIR) return process.env.RAYLAB_CONFIG_DIR;
  if (process.platform === 'darwin') return path.join(os.homedir(), 'Library', 'Application Support', 'raylab-cluster-manager');
  if (process.platform === 'win32') return path.join(process.env.APPDATA || path.join(os.homedir(), 'AppData', 'Roaming'), 'raylab-cluster-manager');
  return path.join(process.env.XDG_CONFIG_HOME || path.join(os.homedir(), '.config'), 'raylab-cluster-manager');
}

function configPath() {
  return path.join(configDir(), 'config.json');
}

// ─── Default config ───────────────────────────────────────────────────────────

const DEFAULT_CONFIG = {
  app_mode: 'unconfigured',
  coordinator: {
    head_host: '127.0.0.1',
    dashboard_host: '127.0.0.1',
    node_ip_address: '',
    ray_port: 6379,
    dashboard_port: 8265,
    client_port: 10001,
    node_manager_port: 18076,
    object_manager_port: 18077,
    preflight_port: 18075,
    cluster_token_ref: 'raylab.cluster_token',
    dashboard_token_ref: 'raylab.dashboard_token',
    bind_private_only: true,
    allow_external_workers: false,
  },
  node_policy: {
    master_enabled: false,
    manual_override: 'auto',
    schedule_enabled: false,
    schedule_windows: [],
    idle_only_enabled: false,
    idle_minutes: 10,
    max_cpu_percent_for_idle: 20,
    max_gpu_percent_for_idle: 10,
  },
  resource_caps: { cpus: 4, gpus: 1, memory_gb: 16, gpu_memory_gb: 12, max_concurrent_jobs: 1 },
  privacy: {
    worker_account: 'raylab-worker',
    worker_account_required: true,
    allow_home_access: false,
    require_runtime_working_dir: true,
    container_runtime: 'docker',
    require_gpu_container_runtime: true,
  },
  object_store: { endpoint_url: '', bucket: '', region: '', access_key_ref: 'raylab.object_store_access_key', secret_key_ref: 'raylab.object_store_secret_key' },
  submitters: [],
  audit: [],
};

// ─── ConfigStore ──────────────────────────────────────────────────────────────

// Simple async mutex — serialise all writes through a promise chain.
let _writeLock = Promise.resolve();

function withWriteLock(fn) {
  _writeLock = _writeLock.then(fn).catch(fn);
  return _writeLock;
}

function load(filePath) {
  const p = filePath || configPath();
  if (!fs.existsSync(p)) {
    return JSON.parse(JSON.stringify(DEFAULT_CONFIG));
  }
  try {
    const raw = fs.readFileSync(p, 'utf8');
    const data = JSON.parse(raw);
    // Deep-merge with defaults so new fields are always present.
    return normalizeConfig(deepMerge(JSON.parse(JSON.stringify(DEFAULT_CONFIG)), data));
  } catch {
    return JSON.parse(JSON.stringify(DEFAULT_CONFIG));
  }
}

function normalizeConfig(config) {
  const coord = config.coordinator || {};

  // Migrate the previous RayLab defaults. Those ports overlapped with too many
  // common local development services and were changed before public rollout.
  if (coord.node_manager_port === 8076) coord.node_manager_port = DEFAULT_CONFIG.coordinator.node_manager_port;
  if (coord.object_manager_port === 8077) coord.object_manager_port = DEFAULT_CONFIG.coordinator.object_manager_port;

  return config;
}

function save(config, filePath) {
  const p = filePath || configPath();
  return withWriteLock(() => {
    fs.mkdirSync(path.dirname(p), { recursive: true });
    fs.writeFileSync(p, JSON.stringify(config, null, 2), 'utf8');
    return config;
  });
}

function appendAudit(event, filePath) {
  return withWriteLock(() => {
    const p = filePath || configPath();
    const config = load(p);
    config.audit = [event, ...(config.audit || [])].slice(0, 1000);
    fs.mkdirSync(path.dirname(p), { recursive: true });
    fs.writeFileSync(p, JSON.stringify(config, null, 2), 'utf8');
    return config;
  });
}

function makeAuditEvent(eventType, message, metadata = {}) {
  return {
    id: crypto.randomUUID(),
    timestamp: new Date().toISOString(),
    event_type: eventType,
    actor: 'system',
    message,
    metadata,
  };
}

// ─── SecretStore ──────────────────────────────────────────────────────────────

const KEYRING_SERVICE = 'RayLabClusterManager';

function devSecretsPath() {
  return path.join(configDir(), 'dev-secrets.json');
}

function _readDevSecrets() {
  try {
    if (fs.existsSync(devSecretsPath())) {
      return JSON.parse(fs.readFileSync(devSecretsPath(), 'utf8'));
    }
  } catch (_) {}
  return {};
}

function _writeDevSecrets(secrets) {
  fs.mkdirSync(configDir(), { recursive: true });
  fs.writeFileSync(devSecretsPath(), JSON.stringify(secrets, null, 2), 'utf8');
}

async function getSecret(ref) {
  try {
    const keytar = require('keytar');
    const val = await keytar.getPassword(KEYRING_SERVICE, ref);
    if (val !== null) return val;
  } catch (_) {}
  return _readDevSecrets()[ref] ?? null;
}

async function setSecret(ref, value) {
  try {
    const keytar = require('keytar');
    await keytar.setPassword(KEYRING_SERVICE, ref, value);
    return;
  } catch (_) {}
  const secrets = _readDevSecrets();
  secrets[ref] = value;
  _writeDevSecrets(secrets);
}

async function getOrCreateToken(ref, length = 48) {
  const existing = await getSecret(ref);
  if (existing) return existing;
  const token = crypto.randomBytes(length).toString('base64url').slice(0, length);
  await setSecret(ref, token);
  return token;
}

// ─── Utilities ────────────────────────────────────────────────────────────────

function deepMerge(target, source) {
  for (const key of Object.keys(source)) {
    if (source[key] && typeof source[key] === 'object' && !Array.isArray(source[key])) {
      if (!target[key] || typeof target[key] !== 'object') target[key] = {};
      deepMerge(target[key], source[key]);
    } else {
      target[key] = source[key];
    }
  }
  return target;
}

module.exports = {
  configDir,
  configPath,
  load,
  save,
  appendAudit,
  makeAuditEvent,
  getSecret,
  setSecret,
  getOrCreateToken,
  DEFAULT_CONFIG,
};
